### PR TITLE
editor: fix issues

### LINKED ISSUE #242: Object data are not cleaned correctly when the property is removed
**Describe the issue**
When a document is saved with a required property, when we edit the document and the property is no more required and removed from editor, some data are kept when saving the document.

**To Reproduce**
1. Go to the document editor on sonardev.test.rero.ch
1. Create a document with type `Preprint` and fill `provisionActivity` field with a start date.
1. Edit the document and change the document type for `Conference poster`.
1. Remove the `provisionActivity` property from editor.
1. Try to save the document, a bad request is thrown from server.

**Expected behavior**
When the data are posted, they must not contain the `provisionActivity` field.

**Screenshots**
provision_activity

### LINKED ISSUE #248: Additional border is displayed for object types
**Describe the issue**
Following this property example:
```json
"editionStatement": {
    "title": "Edition",
    "type": "object",
    "additionalProperties": false,
    "properties": {
        "editionDesignation": {
            "title": "Designation",
            "type": "object",
            "properties": {
                "value": {
                    "title": "Value",
                    "type": "string",
                    "minLength": 1
                }
            },
            "required": [
                "value"
            ]
        },
        "responsibility": {
            "title": "Responsibility",
            "type": "object",
            "properties": {
                "value": {
                    "title": "Value",
                    "type": "string",
                    "minLength": 1
                }
            },
            "required": [
                "value"
            ]
        }
    },
    "required": [
        "editionDesignation"
    ],
    "form": {
        "hide": true,
        "navigation": {
            "essential": true
        },
        "templateOptions": {
            "cssClass": "editor-title"
        }
    }
}
```
The `editor-title` CSS class is added twice in the HTML, causing an extra border in the property.

**To Reproduce**
1. Go to SONAR document's editor https://sonardev.test.rero.ch/manage/records/documents/new.
1. Add the `Edition` field.
1. A border is surrounding the `Edition` properties.

**Expected behavior**
The additional border must not be displayed. In other word, the `editor-title` class must be added only one time.

**Screenshots**
Capture d’écran 2020-08-31 à 13 29 41

### LINKED ISSUE #325: Read-only properties are changeable in editor
**Describe the issue**
If an enum field is marked as read-only, the value can still be changed.

**To Reproduce**
Add this kind of property:
```json
"readonly_changeable": {
    "title": "[issue] Readonly value changeable",
    "description": "Readonly values must not be changeable.",
    "type": "string",
    "default": "Value 1",
    "readOnly": true,
    "enum": [
        "Value 1",
        "Value 2"
    ]
}
```

### LINKED ISSUE #327: Hidden sub-property with hide expression
**Describe the issue**
Hidden sub-property with hide expression cannot be added to the form.

**To Reproduce**
Check this kind of property:
```json
"hidden_sub_property_hide_expr": {
    "title": "[issue] Hidden sub-property with hide expression.",
    "description": "Hidden sub-property with hide expression can be added to the form.",
    "type": "object",
    "additionalProperties": false,
    "properties": {
        "statement": {
            "title": "Statements",
            "type": "array",
            "minItems": 1,
            "items": {
                "title": "Statement",
                "type": "string",
                "minLength": 1
            },
            "form": {
                "hideExpresssion": "true"
            }
        }
    },
    "propertiesOrder": [
        "statement"
    ]
}
```

### LINKED ISSUE #328: Property with hide expression cannot be added
**Describe the issue**
Property with hide expression cannot be added, nor appear in essentials.

**To Reproduce**
Add this kind of property:
```json
"hide_expression_cannot_be_added": {
    "title": "[issue] Property with hide expression.",
    "description": "Property with hide expression cannot be added.",
    "type": "string",
    "minLength": 1,
    "form": {
        "hideExpression": "true",
        "navigation": {
            "essential": true
        }
    }
}
```

### PR DESCRIPTION
**Pull request by jma**
* Fixes tester document search routing.
* Removes useless log messages.
* Fixes import paths.
* Replaces the css classes such as `editor-title` by a `card` wrapper.
* Moves some common code lines in the editor service.
* Moves the field label html code into a specific component.
* Adds new `containerCSSClass`, `itemCSSClass` and `cssClass` to allow
  field grid positions.
* Renders the hide/show/clone button in the children field instead of
  the parent (array, object) component.
* Adds external link support for remote typeahead editor component.
* Closes #328.
* Closes #327.
* Closes #325.
* Closes #248.
* Closes #242.
* Closes rero/rero-ils#1604.
* Closes rero/rero-ils#1601.

Co-Authored-by: Johnny Mariéthoz <Johnny.Mariethoz@rero.ch>

## Why are you opening this PR?

- To solve several issue on SONAR.

## How to test?

-Start the demo server and go to `http://localhost:4200/editor`.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts [hunk 1]
```diff
@@ -15,26 +15,94 @@
...
+    return false;
+  }
+  /**
+   * Add a new element
+   * @param field - FormlyFieldConfig, the field to hide
+   */
+  add() {
+    if (this.field.parent.type === 'array') {
+      const i = this.getIndex() + 1;
+      console.log('i:', i);
```
Comment (sebastiendeleze) [inline projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts:104]: Remove console log

**Review by sebastiendeleze — CHANGES_REQUESTED** [review]
There's a typo in commit message: `routing` instead of `rooting`.

And you can reference the solved issues to close them automatically.

**Comment by sebastiendeleze** [comment]
I tested the issues and some of them seem not to be corrected:

* #248 —> OK
* #196 —> OK
* #325 —> KO (but issue open in ngx-formly)
* #327 —> KO (Work with `hide` but not with `hideExpression`)
* #328 —> KO (Still cannot be added)
* #329 —> KO (Hidden from form but not available in search and listed in `jump to` section)

File: projects/rero/ng-core/src/lib/record/editor/editor.component.scss [hunk 2]
```diff
@@ -17,6 +17,18 @@
...
+@each $breakpoint in map-keys($grid-breakpoints) {
+  @include media-breakpoint-up($breakpoint) {
+    $infix: breakpoint-infix($breakpoint, $grid-breakpoints);
+    @each $prop, $abbrev in (width: w, height: h) {
+      @each $size, $length in $sizes {
+        .#{$abbrev}#{$infix}-#{$size} { #{$prop}: $length !important; }
+      }
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/rero/ng-core/src/lib/record/editor/editor.component.scss:32]: an after that, @jma tells me that he isn't a CSS king .... What's the truth ?

File: projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts [hunk 3]
```diff
@@ -94,9 +94,48 @@ export class EditorService {
...
+   */
+  canHide(field: FormlyFieldConfig) {
+    if (!field.templateOptions.longMode) {
+      return false;
+    }
+    return (
+      !field.templateOptions.required &&
+      !field.hide &&
+      field.hideExpression == null
+    );
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts:116]: I like (me and myself) ternary instruction. In this case 
```
return (!field.templateOptions.longMode)
    ? false
    : ( !field.templateOptions.required &&
        !field.hide &&
        field.hideExpression == null
      );
```

File: projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts [hunk 4]
```diff
@@ -94,9 +94,48 @@ export class EditorService {
...
+  /**
+   * Am I at the root of the form?
+   * @param field - FormlyFieldConfig, the field to hide
+   * @returns boolean, true if I'm the root
+   */
+  isRoot(field: FormlyFieldConfig) {
+    if (!field) {
+      return false;
+    }
+    return field.templateOptions && field.templateOptions.isRoot && field.templateOptions.isRoot === true;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts:128]: ternary is cosmetic ;-)
```
return (!field)
    ? false
    : field.templateOptions && field.templateOptions.isRoot && field.templateOptions.isRoot === true;
```

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.html [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        </ng-container>
+      </ng-core-editor-dropdown-label-editor>
+
+    </div>
+  </ng-template>
+
+  <!-- section title -->
+  <ng-template #title let-f="f">
+    <div class="header">
+      <label class="mb-0" *ngIf="f.templateOptions.label" [attr.for]="f.id" [tooltip]="f.templateOptions.description">
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.html:66]: maybe introduce the possibility to add classes on label ? 
See https://github.com/rero/ng-core/pull/339/files#diff-869d70bf847d429975de17d5bba4a3b9d16c3ce727a971e147a93deaaa8d0fcdR25

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    private _editorService: EditorService,
+    private _translateService: TranslateService
+  ) { }
+
+  /**
+   * Is the dropdown menu displayed?
+   * @param field - FormlyFieldConfig, the correspondig form field config
+   * @returns boolean, true if the menu should be displayed
+   */
+  hasMenu(field: FormlyFieldConfig) {
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts:46]: ternary ?

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  /** Get the current field index position in the parent array.
+   *
+   * @returns number - the index position in the parent array, null if the parent
+   *                   is not an array.
+   */
+  getIndex() {
+    if (this.field.parent.type === 'array') {
+      return Number(this.field.key);
+    }
+    return null;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts:65]: This one is the perfect one for a ternary return : 
``` return (this.field.parent.type === 'array') ? Number(this.field.key) : null; ```

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  /**
+   * Hide the field
+   * @param field - FormlyFieldConfig, the field to hide
+   */
+  remove() {
+    if (this.field.parent.type === 'object') {
+      return this._editorService.hide(this.field);
+    }
+    if (this.field.parent.type === 'array') {
+
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts:129]: blank line

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+   * @returns boolean, true if the field can be hidden
+   */
+  canRemove() {
+    if (this.field.parent.type === 'object') {
+      return this._editorService.canHide(this.field);
+    }
+    if (this.field.parent.type === 'array') {
+      return this.field.parent.templateOptions.canRemove();
+    }
+    return false;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts:153]: maybe a switch statement could be interessant ?
```
switch (this.field.parent.type) { 
    case 'object':  return return this._editorService.canHide(this.field); 
    case 'array': return this.field.parent.templateOptions.canRemove();
    default: return false;
}
```

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+  /**
+   * Is the field can be hidden?
+   * @returns boolean, true if the field can be hidden
+   */
+  canAdd() {
+    if (this.field.parent.type === 'array') {
+      return this.field.parent.templateOptions.canAdd();
+    }
+    return false;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts:164]: ternary ?

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  /**
+   * Add a new element
+   * @param field - FormlyFieldConfig, the field to hide
+   */
+  add() {
+    if (this.field.parent.type === 'array') {
+      const i = this.getIndex() + 1;
+      return this.field.parent.templateOptions.add(this.getIndex() + 1);
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.ts:174]: Either use `i`, either delete line 173

File: projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts [hunk 12]
```diff
@@ -15,26 +15,97 @@
...
+   * @returns boolean, true if the field can be hidden
+   */
+  canRemove() {
+    if (this.field.parent.type === 'object') {
+      return this._editorService.canHide(this.field);
+    }
+    if (this.field.parent.type === 'array') {
+      return this.field.parent.templateOptions.canRemove();
+    }
+    return false;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts:88]: switch ?

File: projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts [hunk 13]
```diff
@@ -15,26 +15,97 @@
...
+
+  /**
+   * Is the field can be hidden?
+   * @returns boolean, true if the field can be hidden
+   */
+  canAdd() {
+    if (this.field.parent.type === 'array') {
+      return this.field.parent.templateOptions.canAdd();
+    }
+    return false;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts:99]: ternary again ?

File: projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts [hunk 14]
```diff
@@ -15,26 +15,97 @@
...
+  }
+  /**
+   * Add a new element
+   * @param field - FormlyFieldConfig, the field to hide
+   */
+  add() {
+    if (this.field.parent.type === 'array') {
+      const i = this.getIndex() + 1;
+      return this.field.parent.templateOptions.add(this.getIndex() + 1);
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/wrappers/horizontal-wrapper/horizontal-wrapper.component.ts:108]: use `i` or delete the line 107

**Review by zannkukai — CHANGES_REQUESTED** [review]
Most of comments are cosmetics ;-) you can skip them

File: projects/rero/ng-core/src/lib/record/editor/editor.component.scss (hunk 2 above)
[the remark this answers is not in the document]
Comment (jma) [reply projects/rero/ng-core/src/lib/record/editor/editor.component.scss:32]: It is just a copy/paste from the web.

File: projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts (hunk 3 above)
Comment (jma) [reply projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts:116]: Seem to me more difficult to read. @sebastiendeleze any advises?

File: projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.html (hunk 5 above)
Comment (jma) [reply projects/rero/ng-core/src/lib/record/editor/widgets/label/label.component.html:66]: I prefer the more generic approach: https://github.com/rero/ng-core/pull/338/files#diff-869d70bf847d429975de17d5bba4a3b9d16c3ce727a971e147a93deaaa8d0fcdR24

File: projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts (hunk 3 above)
Comment (sebastiendeleze) [reply projects/rero/ng-core/src/lib/record/editor/services/editor.service.ts:116]: +1 for use ternary, but only if expression is quite simple. In this case, it is difficult to read in my opinion. But I would suggest to merge statements like this:
```
return (
      field.templateOptions.longMode &&
      !field.templateOptions.required &&
      !field.hide &&
      field.hideExpression == null
    );
```

**Review by zannkukai — APPROVED** [review]

**Comment by sebastiendeleze** [comment]
When I tried the editor on SONAR, the main field wrapper and labels are appearing twice. I noticed this problem in ng-core-tester, too.

Capture d’écran 2021-01-19 à 14 15 54

**Comment by sebastiendeleze** [comment]
When I try to create a new document, some fields appear both in the left side and in the editor.

collection

**Comment by sebastiendeleze** [comment]
Based on the schema property here https://github.com/rero/sonar/blob/staging/sonar/modules/documents/jsonschemas/documents/document-v1.0.0_src.json#L313, I think the subtitles must not be present in the editor by default:

Capture d’écran 2021-01-19 à 14 44 22

**Comment by jma** [comment]
> When I tried the editor on SONAR, the main field wrapper and labels are appearing twice. I noticed this problem in ng-core-tester, too.
> 
> Capture d’écran 2021-01-19 à 14 15 54

It should be solved

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

File: projects/rero/ng-core/src/lib/record/editor/type/multischema/multischema.component.ts [hunk 15]
```diff
@@ -14,14 +14,14 @@
  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
-import { Component } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
```
Comment (sebastiendeleze) [inline projects/rero/ng-core/src/lib/record/editor/type/multischema/multischema.component.ts:17]: On Init is not used.

**Review by sebastiendeleze — APPROVED** [review]

**Pull request merged**